client.go

```diff
@@ -9,257 +9,118 @@ import (
 	"net"
 	"strconv"
 	"strings"
-	"sync"
 	"time"
-
-	tomb "gopkg.in/tomb.v2"
 )
 
 type Client struct {
+	// Logger for messages. By defaut this will be a NilLogger
 	Logger Logger
 
-	handler     Handler
-	connected   bool
+	// Internal things
 	currentNick string
-	nick        string
-	user        string
-	name        string
-	password    string
-	lock        *sync.Mutex
 	conn        io.ReadWriteCloser
-	t           *tomb.Tomb
-	write       chan string
+	in          *bufio.Reader
 }
 
-func NewClient(handler Handler, nick string, user string, name string, pass string) *Client {
-	// Create the client
-	c := &Client{
-		nil,
-		handler,
-		false,
-		nick,
-		nick,
-		user,
-		name,
-		pass,
-		&sync.Mutex{},
-		nil,
-		&tomb.Tomb{},
-		make(chan string),
+func Dial(addr string, nick, user, name, pass string) (*Client, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
 	}
 
-	return c
-}
-
-func (c *Client) CurrentNick() string {
-	return c.currentNick
+	return NewClient(conn, nick, user, name, pass), nil
 }
 
-func (c *Client) Dial(host string) error {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	if c.connected {
-		return errors.New("Already connected")
-	}
-
-	var err error
-	c.conn, err = net.Dial("tcp", host)
+func DialTLS(addr string, c *tls.Config, nick, user, name, pass string) (*Client, error) {
+	conn, err := tls.Dial("tcp", addr, c)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	c.connected = true
-
-	return c.start()
+	return NewClient(conn, nick, user, name, pass), nil
 }
 
-func (c *Client) DialTLS(host string, conf *tls.Config) error {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	if c.connected {
-		return errors.New("Already connected")
+func NewClient(rwc io.ReadWriteCloser, nick, user, name, pass string) *Client {
+	// Create the client
+	c := &Client{
+		&NilLogger{},
+		nick,
+		rwc,
+		bufio.NewReader(rwc),
 	}
 
-	var err error
-	c.conn, err = tls.Dial("tcp", host, conf)
-	if err != nil {
-		return err
+	// Send the info we need to
+	if len(pass) > 0 {
+		c.Writef("PASS %s", pass)
 	}
 
-	c.connected = true
+	c.Writef("NICK %s", nick)
+	c.Writef("USER %s 0.0.0.0 0.0.0.0 :%s", user, name)
 
-	return c.start()
+	return c
+}
+
+func (c *Client) CurrentNick() string {
```

Comment:
The idea is that we want it read only. But I suppose if it's documented it would be alright?